src/server/pps/server/api_server.go

```diff
@@ -3959,20 +3959,100 @@ func (a *apiServer) GetProjectDefaults(ctx context.Context, req *pps.GetProjectD
 	return &pps.GetProjectDefaultsResponse{ProjectDefaultsJson: projectDefaults.Json}, nil
 }
 
-// SetProjectDefaults is a stub, for use in testing.
-//
-// TODO(CORE-1712): implement as an RPC.
+// SetProjectDefaults sets the defaults for a project.  If regenerate is true,
+// then the effective spec for each pipeline in the project will be regenerated
+// and if has changed it will be saved; if reprocess is true then regenerated
+// (and changed) pipelines will reprocess previously-processed datums.
 func (a *apiServer) SetProjectDefaults(ctx context.Context, req *pps.SetProjectDefaultsRequest) (*pps.SetProjectDefaultsResponse, error) {
 	if req.Project == nil || req.Project.Name == "" {
 		req.Project = &pfs.Project{Name: pfs.DefaultProjectName}
 	}
+	var cdg = &cachedDefaultsGetter{
+		apiServer:       a,
+		projectDefaults: map[string]string{req.GetProject().String(): req.GetProjectDefaultsJson()},
+	}
+
+	var (
+		pp map[*pps.Pipeline]*pps.CreatePipelineTransaction
+	)
+	cResp, err := a.GetClusterDefaults(ctx, &pps.GetClusterDefaultsRequest{})
+	if err != nil {
+		return nil, unknownError(ctx, "could not get cluster defaults", err)
+	}
+	if _, _, err := makeEffectiveSpec(cResp.GetClusterDefaultsJson(), req.GetProjectDefaultsJson(), `{}`); err != nil {
+		return nil, badRequest(ctx, fmt.Sprintf("could not merge project defaults %s into cluster defaults %s", req.GetProjectDefaultsJson(), cResp.GetClusterDefaultsJson()), []*errdetails.BadRequest_FieldViolation{
+			{Field: "project_defaults_json", Description: err.Error()},
+		})
+	}
+
+	if req.Regenerate {
+		// Determine if the new defaults imply changes to any pipelines.
+		// To do this, each pipeline’s info is synthesized to a
+		// CreatePipelineRequest using the same logic already ultimately
+		// used by `pachctl edit pipeline`, then that is turned into
+		// JSON and merged with the new defaults.  The result of the
+		// merger is then unmarshalled into a CreatePipelineRequest and
+		// equality is checked with proto.Equal.
+		pp = make(map[*pps.Pipeline]*pps.CreatePipelineTransaction)
+		if err := a.listPipeline(ctx, &pps.ListPipelineRequest{Details: true}, func(pi *pps.PipelineInfo) error {
```

Comment:
I don't think ListPipelineRequest.Details is being interpreted anywhere in the system